src/address.rs

```diff
@@ -1,91 +1,338 @@
             .sender
-            .unbounded_send(Box::new(envelope));
-
-        rx.map_err(|_| Disconnected)
+            .unbounded_send(ManagerMessage::Message(Box::new(envelope)));
+        MessageResponseFuture::Result(rx)
     }
 
-    /// Sends a [`Message`](trait.Message.html) that will be handled asynchronously to the actor,
-    /// and waits for a response. If this returns `Err(Disconnected)`, then the actor is stopped
-    /// and not accepting messages.
-    pub fn send_async<M>(&self, message: M) -> impl Future<Output = Result<M::Result, Disconnected>>
+    fn send_async<M>(&self, message: M) -> MessageResponseFuture<M>
     where
         M: Message,
         A: AsyncHandler<M> + Send,
         for<'a> A::Responder<'a>: Future<Output = M::Result> + Send,
     {
         let (envelope, rx) = AsyncReturningEnvelope::new(message);
-
         let _ = self
             .sender
-            .unbounded_send(Box::new(envelope));
+            .unbounded_send(ManagerMessage::Message(Box::new(envelope)));
+        MessageResponseFuture::Result(rx)
+    }
+}
 
-        rx.map_err(|_| Disconnected)
+// Required because #[derive] adds an A: Clone bound
+impl<A: Actor> Clone for Address<A> {
+    fn clone(&self) -> Self {
+        Address {
+            sender: self.sender.clone(),
+            ref_counter: self.ref_counter.clone(),
+        }
     }
 }
 
-/// The actor is no longer running and disconnected
-#[derive(Clone, Eq, PartialEq, Debug)]
-pub struct Disconnected;
+impl<A: Actor> Drop for Address<A> {
+    fn drop(&mut self) {
+        // Context holds one strong address (for `Context::address`) and so does ActorManager, so if
+        // there are 3 strong addresses, this would be the only external one in existence. Therefore, we
+        // should notify the ActorManager that there are potentially no more strong Addresses and the
+        // actor should be stopped.
+        if Arc::strong_count(&self.ref_counter) == 3 {
+            let _ = self.sender.unbounded_send(ManagerMessage::LastAddress);
+        }
+    }
+}
+
+/// A `WeakAddress` is a reference to an actor through which [`Message`s](trait.Message.html) can be
+/// sent. It can be cloned. Unlike [`Address`](struct.Address.html), a `WeakAddress` will not inhibit
+/// the dropping of an actor. It is created by the [`Address::weak`](struct.Address.html) method.
+pub struct WeakAddress<A: Actor> {
+    pub(crate) sender: UnboundedSender<ManagerMessage<A>>,
+    ref_counter: Weak<()>,
+}
+
+impl<A: Actor> AddressExt<A> for WeakAddress<A> {
+    fn do_send<M>(&self, message: M) -> Result<(), Disconnected>
+    where
+        M: Message,
+        A: Handler<M> + Send,
+    {
+        // Check that there are external strong addresses. If there are none, the actor is
+        // disconnected and our message would interrupt its dropping. strong_count() == 2 because
+        // Context and manager both hold a strong arc to the refcount
+        if self.ref_counter.strong_count() > 2 {
```

Comment:
I don't know if it can be avoided but it is a bit confusing to have 2 here be the minimum count. Maybe even just an internal function to get the actual number of strong counts (`ref_counter.strong_count() - 2`) would be nice

Reply:
Internal function would have to be in quite a few places but overall might save on comments I guess